src/algorithms/peter_norving.py

```diff
@@ -0,0 +1,165 @@
+        """
+
+        return [key + value for key in row_key_list for value in column_key_list]
+
+    def parse_grid(self, grid):
+        """
+        Converts grid game string into a dictionary of possible values,
+        {square: digits}, or return False if a contradiction is detected.
+
+        Keyword arguments:
+        grid -- String that contains the sudoku game, i.e:
+                3000800000007000051000000000000003600020 .......
+        returns a dictionary with all the possible values per square. i.e:
+               {'I6': '1379', 'H9': '679', 'I2': '12369' ........
+        """
+
+        values = dict((square, self.digits) for square in self.squares)
+        for (square, digit) in list(self.create_game_dict(grid).items()):
+            if digit in self.digits and not self.assign(values, square,
+                    digit):
+                return False
+        return values
+
+    def create_game_dict(self, grid):
+        """
+        Convert grid game string into a dictionary  of {square: char}
+        with '0' for empties.
+
+        Keyword arguments:
+        grid -- String that contains the sudoku game, i.e:
+                3000800000007000051000000000000003600020 .......
+        returns grid dict with the filled values ready to play.
+        """
+
+        chars = [colum for colum in grid if colum in self.digits
+                 or colum in '0']
+        assert len(chars) == 81
+        return dict(list(zip(self.squares, chars)))
+
+    def assign(self, values, square, digit):
+        """
+        Eliminate the list of "other_values"(possible solutions per square)
+        except the sent digit from values[square] and then propagate.
+
+        Keyword arguments:
+        values -- initial dictionary with values.
+        square -- square positions.
+        digit --  selected digit.
+        returns updated solution value list
+        or return False if a contradiction is detected.
+        """
+
+        other_values = values[square].replace(digit, '')
+        if all(self.eliminate(values, square, temp_digit)
+               for temp_digit in other_values):
+            return values
+        else:
+            return False
+
+    def eliminate(self, values, square, digit):
+        """
+        Eliminate 'digit' from values[square]. Propagate when values or places <= 2.
+
+        Keyword arguments:
+        values -- initial dictionary with values.
+        square -- square positions
+        digit --  selected digit.
+        returns updated solution value list
+        or return False if a contradiction is detected.
+        """
+
+        if digit not in values[square]:
+            return values
+        values[square] = values[square].replace(digit, '')
+        if len(values[square]) == 0:
+            return False
+        elif len(values[square]) == 1:
+            temp_digit = values[square]
+            if not all(self.eliminate(values, temp_square, temp_digit)
```

Comment:
don't split into two lines this kind of scenarios and use the normal way of the for statement, I think you use the inline way just to avoid the nested levels, you need to improve this